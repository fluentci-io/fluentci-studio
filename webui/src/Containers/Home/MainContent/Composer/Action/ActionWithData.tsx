import { FC } from "react";
import Action from "./Action";
import { useRecoilState } from "recoil";
import { ComposerState } from "../ComposerState";
import { Pipeline } from "../NewActionModal/NewActionModalWithData";

export type ActionWithDataProps = {
  action: Pipeline;
  index: number;
  onClickAction: (action: Pipeline, index: number) => void;
  onDelete: (position: number) => void;
  onDuplicate: (position: number) => void;
};

const ActionWithData: FC<ActionWithDataProps> = (props) => {
  const [actions, setActions] = useRecoilState(ComposerState);
  const activate = (checked: boolean) => {
    const updated = [...actions];
    updated[props.index] = {
      ...updated[props.index],
      active: checked,
    };
    setActions(updated);
  };
  return <Action {...props} activate={activate} />;
};

export default ActionWithData;
