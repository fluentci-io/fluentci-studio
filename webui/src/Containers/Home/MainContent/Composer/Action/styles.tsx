import { css } from "@emotion/react";
import styled from "@emotion/styled";

export const Animated = styled.div`
  -webkit-animation: zoom-in-zoom-out 500ms; /* Safari, Chrome and Opera > 12.1 */
  -moz-animation: zoom-in-zoom-out 500ms; /* Firefox < 16 */
  -ms-animation: zoom-in-zoom-out 500ms; /* Internet Explorer */
  -o-animation: zoom-in-zoom-out 500ms; /* Opera < 12.1 */
  animation: zoom-in-zoom-out 500ms;

  @keyframes zoom-in-zoom-out {
    0% {
      transform: scale(1, 1);
    }
    50% {
      transform: scale(1.5, 1.5);
    }
    100% {
      transform: scale(1, 1);
    }
  }
`;

export const Container = styled(Animated)<{ active?: boolean }>`
  display: flex;
  flex-direction: row;
  align-items: center;
  border-radius: 8px;
  padding: 10px;
  border: 1px solid #5324ffa3;
  height: 78px;
  box-shadow: 2px 4px #5324ff20;
  padding-left: 18px;
  padding-right: 18px;
  cursor: pointer;
  ${(props) =>
    !props.active &&
    css`
      opacity: 0.5;
    `}
`;

export const ActionName = styled.div`
  color: #fff;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const PopoverButton = styled.button`
  width: 40px;
  height: 40px;
  cursor: pointer;
  background-color: initial;
  border: none;
  color: #fff;
`;

export const Popover = styled.div`
  background-color: #0f0124;
  color: #fff;
`;
