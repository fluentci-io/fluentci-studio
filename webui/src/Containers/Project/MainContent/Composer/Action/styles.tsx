import { css } from "@emotion/react";
import styled from "@emotion/styled";

export const Container = styled.div<{ active?: boolean }>`
  display: flex;
  flex-direction: row;
  align-items: center;
  border-radius: 8px;
  padding: 10px;
  border: 1px solid #5324ffa3;
  height: 78px;
  box-shadow: 2px 4px #5324ff20;
  padding-left: 18px;
  padding-right: 18px;
  cursor: pointer;
  ${(props) =>
    !props.active &&
    css`
      opacity: 0.5;
    `}
`;

export const ActionName = styled.div`
  color: #fff;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;

export const PopoverButton = styled.button`
  width: 40px;
  height: 40px;
  cursor: pointer;
  background-color: initial;
  border: none;
  color: #fff;
`;

export const Popover = styled.div`
  background-color: #0f0124;
  color: #fff;
`;
