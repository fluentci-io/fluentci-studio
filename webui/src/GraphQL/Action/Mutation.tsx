import { gql } from "@apollo/client";
import { ActionFragment } from "../Fragment";

export const SAVE_ACTIONS = gql`
  mutation SaveActions($projectId: ID!, $actions: [ActionInput!]!) {
    saveActions(projectId: $projectId, actions: $actions) {
      ...ActionFragment
    }
  }
  ${ActionFragment}
`;
