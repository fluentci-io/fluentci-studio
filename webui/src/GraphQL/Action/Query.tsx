import { gql } from "@apollo/client";
import { ActionFragment } from "../Fragment";

export const GET_ACTIONS = gql`
  query GetActions($projectId: ID!) {
    actions(projectId: $projectId) {
      ...ActionFragment
    }
  }
  ${ActionFragment}
`;

export const EXPORT_ACTIONS = gql`
  query ExportActions($projectId: ID!, $plateform: String!) {
    exportActions(projectId: $projectId, plateform: $plateform)
  }
`;
