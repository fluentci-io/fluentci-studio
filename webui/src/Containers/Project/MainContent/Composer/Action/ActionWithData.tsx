import { FC } from "react";
import Action from "./Action";
import { useRecoilState } from "recoil";
import { ComposerState } from "../ComposerState";
import { Pipeline } from "../NewActionModal/NewActionModalWithData";
import { useSaveActionsMutation } from "../../../../../Hooks/GraphQL";
import { useParams } from "react-router-dom";

export type ActionWithDataProps = {
  action: Pipeline;
  index: number;
  onClickAction: (action: Pipeline, index: number) => void;
  onDelete: (position: number) => void;
  onDuplicate: (position: number) => void;
};

const ActionWithData: FC<ActionWithDataProps> = (props) => {
  const { id } = useParams();
  const [saveActions] = useSaveActionsMutation();
  const [actions, setActions] = useRecoilState(ComposerState);
  const activate = (checked: boolean) => {
    const updated = [...actions];
    updated[props.index] = {
      ...updated[props.index],
      active: checked,
    };
    _setActions(updated);
  };

  const _setActions = (actions: Pipeline[]) => {
    setActions(actions);
    saveActions({
      variables: {
        projectId: id!,
        actions: actions.map((action) => ({
          commands: action.command,
          enabled: !!action.active,
          logo: action.logo,
          name: action.actionName!,
          plugin: action.name!,
          useWasm: !!action.useWasmPlugin,
          githubUrl: action.githubUrl,
        })),
      },
    });
  };

  return <Action {...props} activate={activate} />;
};

export default ActionWithData;
