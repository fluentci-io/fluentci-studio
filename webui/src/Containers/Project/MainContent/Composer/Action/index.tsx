import Action from "./ActionWithData";

export default Action;
