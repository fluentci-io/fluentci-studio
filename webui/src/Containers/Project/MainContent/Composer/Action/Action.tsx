import { ChangeEvent, FC } from "react";
import { Terminal } from "@styled-icons/bootstrap";
import { EllipsisVertical } from "@styled-icons/fa-solid";
import { StatefulPopover } from "baseui/popover";
import { Checkbox, STYLE_TYPE } from "baseui/checkbox";
import { Container, ActionName, PopoverButton } from "./styles";
import { Pipeline } from "../NewActionModal/NewActionModalWithData";
import { StatefulMenu } from "baseui/menu";

export type ActionProps = {
  action: Pipeline;
  index: number;
  onClickAction: (action: Pipeline, index: number) => void;
  onDelete: (position: number) => void;
  onDuplicate: (position: number) => void;
  activate: (checked: boolean) => void;
  disabled: boolean;
};

const Action: FC<ActionProps> = (props) => {
  const { action, index, onClickAction, onDelete, onDuplicate, activate } =
    props;
  return (
    <Container
      active={action.active}
      className="action"
      onClick={(e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        if (!(e.target as HTMLElement).className) {
          return;
        }
        if ((e.target as HTMLElement).className) {
          if (typeof (e.target as HTMLElement).className === "object") {
            return;
          }
          if (
            !(e.target as HTMLElement).className.includes("action") &&
            !(e.target as HTMLElement).className.includes("css")
          ) {
            return;
          }
        }
        onClickAction(action, index ? 1 : index);
      }}
    >
      {action.logo && (
        <img
          className="action"
          src={action.logo}
          alt={action.name}
          style={{
            backgroundColor:
              action.name.startsWith("deno") ||
              action.name.startsWith("github") ||
              action.name.startsWith("rust") ||
              action.name.startsWith("trivy") ||
              action.name.startsWith("symfony") ||
              action.name.startsWith("flakestry") ||
              action.name.startsWith("symfony") ||
              action.name.startsWith("heroku") ||
              action.name.startsWith("django") ||
              action.name.startsWith("terraform") ||
              action.name.startsWith("prisma")
                ? "#fff"
                : "initial",
            maxWidth: 34,
            borderRadius: 2,
            padding: 5,
            marginRight: 15,
          }}
        />
      )}
      {!action.logo && (
        <Terminal
          size={32}
          color="#fff"
          style={{ marginLeft: 4, marginRight: 15 }}
        />
      )}
      <div
        className="action"
        style={{ display: "flex", flex: 1, overflow: "hidden" }}
      >
        <ActionName>
          {action.actionName || action.command?.split("\n")?.reverse()[0]}
        </ActionName>
      </div>
      <Checkbox
        disabled={props.disabled}
        checked={action?.active}
        onChange={(e: ChangeEvent<HTMLInputElement>) => {
          activate(e.currentTarget.checked);
        }}
        checkmarkType={STYLE_TYPE.toggle_round}
        overrides={{
          Toggle: {
            style: {
              backgroundColor: "#fff",
              opacity: props.disabled ? 0.5 : 1,
            },
          },
          ToggleTrack: {
            style: {
              backgroundColor: action?.active ? "#441ecdfe" : "#451ecd52",
            },
          },
          ToggleInner: {
            style: {
              backgroundColor: "#fff",
            },
          },
          Checkmark: {
            style: {
              backgroundColor: "#fff",
            },
          },
        }}
      ></Checkbox>

      <StatefulPopover
        placement="bottomRight"
        content={({ close }) => (
          <StatefulMenu
            items={[
              { label: <span style={{ color: "#fff" }}>Duplicate</span> },
              {
                label: <span style={{ color: "#ff0094" }}>Delete action</span>,
              },
            ]}
            overrides={{
              List: {
                style: {
                  border: "1px solid rgb(71 5 94 / 29%)",
                },
              },
              Option: {
                props: {
                  overrides: {
                    ListItem: {
                      style: ({
                        $theme,
                      }: {
                        $theme: { primaryFontFamily: string };
                      }) => ({
                        fontFamily: $theme.primaryFontFamily,
                      }),
                    },
                  },
                },
              },
            }}
            onItemSelect={({
              item,
            }: {
              item: { label: { props: { children: string } } };
            }) => {
              switch (item.label.props.children) {
                case "Delete action":
                  onDelete(index);
                  break;
                case "Duplicate":
                  onDuplicate(index);
                  break;
                default:
                  break;
              }
              close();
            }}
          />
        )}
        accessibilityType={"tooltip"}
      >
        <PopoverButton
          onClick={(e) => e.stopPropagation()}
          disabled={props.disabled}
        >
          <EllipsisVertical size={20} />
        </PopoverButton>
      </StatefulPopover>
    </Container>
  );
};

export default Action;
