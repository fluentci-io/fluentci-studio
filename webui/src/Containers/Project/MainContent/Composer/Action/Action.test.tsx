import { render } from "@testing-library/react";
import Action from "./Action";

describe("Action", () => {
  it("should render the component", () => {
    const { container } = render(
      <Action
        action={{
          id: "cfah1yaym5ol34wyoqyov75r",
          name: "go",
          actionName: "tests",
          active: true,
          useWasmPlugin: true,
          command: "test ./...",
          description: "",
          githubUrl: "https://github.com/fluent-ci-templates/go-pipeline",
          logo: "https://cdn.jsdelivr.net/gh/fluent-ci-templates/.github/assets/go-original-wordmark.svg",
          packageId: "go",
          downloads: 0,
          version: "",
          license: "",
          defaultBranch: "",
          updatedAt: "",
          comingSoon: false,
        }}
        index={1}
        onClickAction={vi.fn()}
        onDelete={vi.fn()}
        onDuplicate={vi.fn()}
        activate={vi.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
