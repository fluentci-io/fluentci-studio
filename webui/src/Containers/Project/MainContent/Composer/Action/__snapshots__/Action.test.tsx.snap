// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Action > should render the component 1`] = `
<div>
  <div
    class="action css-1gebutb"
  >
    <img
      alt="go"
      class="action"
      src="https://cdn.jsdelivr.net/gh/fluent-ci-templates/.github/assets/go-original-wordmark.svg"
      style="max-width: 34px; border-radius: 2px; padding: 5px; margin-right: 15px;"
    />
    <div
      class="action"
      style="display: flex; flex: 1; overflow: hidden;"
    >
      <div
        class="css-3azhy3"
      >
        tests
      </div>
    </div>
    <label
      class=""
      data-baseweb="checkbox"
    >
      <div
        class=""
      >
        <div
          class=""
        />
      </div>
      <input
        checked=""
        class=""
        type="checkbox"
        value=""
      />
    </label>
    <button
      class="css-1mybska"
      id="1__anchor"
    >
      <svg
        aria-hidden="true"
        class="StyledIconBase-sc-ea9ulj-0 jvZCGA"
        fill="currentColor"
        focusable="false"
        height="20"
        viewBox="0 0 128 512"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M64 360c30.9 0 56 25.1 56 56s-25.1 56-56 56-56-25.1-56-56 25.1-56 56-56zm0-160c30.9 0 56 25.1 56 56s-25.1 56-56 56-56-25.1-56-56 25.1-56 56-56zm56-104c0 30.9-25.1 56-56 56S8 126.9 8 96s25.1-56 56-56 56 25.1 56 56z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
